### DISCUSSION TITLE
Using environment variables in custom code

### DISCUSSION THREAD

**Discussion by DonKirkham** [microsoft/PowerPlatformConnectors#3566] [Q&A] [answered]
I am working on a custom connector that uses custom code to execute the actions. This connector is in a Power Platform solution that has several defined Environment Variables. Is it possible to use those environment variables in the custom code? I've explored the Environment class, but that does not seem to be correct. It seems like it might be in the Context object, but I just can't find it. Thanks for any help provided, even if it's to say "stop wasting your time!"

**Comment by troystaylor** [comment]
You can set the default of a query or header parameter to the environment var and then get that parameter using custom code. That’s the only way I know to do it

**Reply by DonKirkham** [reply]
Thanks Troy. After I posted yesterday, I started thinking that query param might work, but didn't have time to test. I like the header param even better since I can strip the parameter before sending. I will set it up and report back.

**Reply by DonKirkham** [reply]
OK, I have tried adding a test environment variable to the header. The value of the "new_test" environment variable is "This is a test"

I used the following swagger:

	parameters:
        - name: test    
          in: header 
          type: string 
          default: '@environmentVariables("new_test")' 
          x-ms-visibility: internal 
          required: true 

Unfortunately, the headers in the "http post" does not pull in the environment variable:

	{
	  "Authorization": "Bearer eyJ0eXA...,
	  "Accept": "*/*",
	  "test": "@environmentVariables(\"new_test\")"
	}

I use an environment variable for the host value and that is working fine.

What am I missing?  Thanks, Don

**Comment by troystaylor** [comment]
Try default: '@environmentVariables('new_test')'

**Reply by DonKirkham** [reply]
That results in a syntax error in the swagger file, as swagger interprets the second tick mark as the end of the string and expects a new line and indent before the next property.
 
![image](https://github.com/user-attachments/assets/a0f0cdd5-bf3d-4d6c-a460-9374a766eee5)
![image](https://github.com/user-attachments/assets/69a63716-9371-408f-908a-1cc75cfc4d47)

**Comment by troystaylor** [comment]
Ah, I forgot that is what YAML does. Here you go:
![brave_VkyHStcYwS](https://github.com/user-attachments/assets/8d45f9d5-de0e-4d8e-9825-7e2611e86bfa)

**Reply by DonKirkham** [reply]
Unfortunately, that didn't work either. I have tried every combo of quotes and double quotes I can think of.  On the Test tab of the connector, it shows the test key as follows
{
  ...
  "Accept": "*/*",
  "test": "@environmentVariables('new_test')"
}

And when I pull the headers in the Code using JsonConvert.SerializeObject(Context.Request.Headers), I get 
{
...
    { "Key": "test", "Value": ["@environmentVariables('new_test')"] },
...
}

**Comment by troystaylor** [comment]
Are you sure the connector is in a solution?

**Reply by DonKirkham** [reply]
Yes. I am using another environment variable to set the host and it works flawlessly. 

Searching finds an article on MS Learn [(https://learn.microsoft.com/en-us/connectors/custom-connectors/environment-variables)](https://learn.microsoft.com/en-us/connectors/custom-connectors/environment-variables) that has the following snippet:

> "You can also use environment variables in solution custom connectors to update key custom connector properties such as Host, Base URL, Client ID, Client Secret, Login Url, and Refresh Url. For a detailed tutorial, you can read [this blog](https://powerautomate.microsoft.com/blog/environment-variables-in-custom-connectors/)."

The link to the blog didn't work for me and search in the blog manually eventually lead to a simplistic post with a link back to the same Learn article.

The article is from 2022 and I was hoping the capabilities had expanded since then.

**Comment by troystaylor** [comment, marked as answer]
Rather than as default on the parameter, try setting it with a policy. It should be like this, even though this uses a connection parameter: 
![image](https://github.com/user-attachments/assets/5efef085-324f-483e-9edb-2666783d7e73)

**Reply by DonKirkham** [reply]
That worked!!!

This code

> Context.Request.Headers.TryGetValues("test", out var testValue);  
> Context.Logger.LogInformation($"test: {testValue.First()}");


results in
> [2024-07-18T21:39:33.4606366+00:00] Information:   {OriginalFormat}: test: This is a test


Thank you so much for you help (and patience!)
